Feedback wanted: Temporary failure error handling

rbt opened the pull request:
This is a WIP needing feedback.

Ideally the framework (Cro, Bailador, etc.) would have a retry loop and be able to catch temporary type errors and do the retry handling itself.

Outstanding issues
---
1.  throws-like expects "Bool::False" but is-temporary only returns "False". This feels like a throws-like problem.
1. Should SQLState be up a level? DBI on perl5 has non-SQL databases so I've opted to apply it to the Pg exception only but every SQL database should be setting the SQLState in a consistent manner. State exceptions are a part of the spec.
1. To determine it's a ::Pg exception a file requires "need DBDish::Pg::ErrorHandling;". Is it possible to eliminate that so the standard "use DBDish;" covers it?

I'm quite new to Perl6. I didn't see anything in the documentation like an X-Temporary Role to decorate exception objects but there is probably a more generic way of achieving this.


Commit message
---
DBs have numerous temporary errors from deadlocks to serialization failures to network failures. SQLState contains the error code (largely standardized across SQL DBs, but not completely) for various situations which may require a retry or allow the client to perform other correctable work.

Export sqlstate and all other available fields for Pg in X::DBDish::DBError::Pg (extends X::DBDish::DBError).

While here, shorten the message to just the error component for Pg. Detail and Hint are available via other fields and is quite noisy (it typically starts with Error: ERROR: as DBIish adds an 'Error:' prefix too).

Add is-temporary which checks if the sqlstate for Pg is a temporary failure and the transaction should typically be retried immediately by the client. These states for Pg include:
  Class 08\w{3}: All connection exceptions (possible network issues)
  40001 serialization_failure
  40P01 deadlock_detected
  Class 57\w{3}: Operator Intervention (early/forced connection termination)
  72000 snapshot_too_old

The intent is that a tool like Cro can check for is-temporary across all DB drivers and restart working on the request with very little assistance from the developer other than perhaps a "sub routes(:restartable)" type markup to indicate their code is safe to re-execute.

MasterDuke17 commented on lib/DBDish/Pg/ErrorHandling.pm6:
```diff
@@ -0,0 +1,38 @@
+        #  - Class 57\w{3}: Operator Intervention (early/forced connection termination)
+        #  - 72000 snapshot_too_old
+        method is-temporary {
+           return True when ($.sqlstate ~~ /^ '08'<[alnum]> ** 3 | '40001'| '40P01' | '57'<[alnum]> ** 3 | '72000' $/ );
```
This may just be a matter of style, but these two lines could also be written as just:
`so $.sqlstate ~~ /^ '08'<[alnum]> ** 3 | '40001'| '40P01' | '57'<[alnum]> ** 3 | '72000' $/`

jonathanstowe commented on lib/DBDish/Pg/StatementHandle.pm6:
```diff
@@ -17,7 +18,33 @@ method !handle-errors {
         self.reset-err;
     } else {
-        self!set-err($!result.PQresultStatus, $!pg_conn.PQerrorMessage);
+        self!error-dispatch: X::DBDish::DBError::Pg.new(
```
It's entirely a matter of personal taste, but seeing as nearly all the attributes are derived from the ```$!result``` I'd be tempted to just pass that to the constructor and de-structure it in the ```BUILD``` or ```TWEAK```` as appropriate.

rbt commented:
I've implemented both changes as recommended (rebased to merge commits), added a bit of documentation, and worked around the lack of Bool support in throws-like tests.

Unless there are other concerns, I think this is ready for merge now.

rbt commented:
Fixed indents. Also setup a version restriction on the tests as EOL'd versions (9.2 and earlier) do not support all fields; Pg 9.3 hits EOL in less than a month.

jonathanstowe commented on lib/DBDish/Pg/ErrorHandling.pm6:
```diff
@@ -0,0 +1,64 @@
+        #  - 40P01 deadlock_detected
+        #  - Class 57\w{3}: Operator Intervention (early/forced connection termination)
+        #  - 72000 snapshot_too_old
+        method is-temporary {
```
I'd be inclined to put an implementation of  ```is-temporary```  in the ```X::DBDish::DBError``` that just returns False in order that all drivers get a similar interface.

rbt replied:
That makes sense.

I've pushed that change along with a comment which should be helpful to other driver authors.
